### PR TITLE
feat(ts/analyzer): intersection reduction

### LINKED ISSUE #377: Use correct error code for readonly properties
https://github.com/dudykr/stc/blob/d50b685dd80aaa0ca4227551bb9cd26af0fd7af4/crates/stc_ts_type_checker/tests/conformance/types/intersection/intersectionTypeReadonly.ts#L17

This should fail with `TS2540`, but currently stc fails with 

```
TS2339

  x context: tried to access property of an object to calculate type of a member expression
  | context: tried to access property of an object (Base#6917;
  | Member as {
  |     readonly value: number;
  | };, id_ctx = Var)
  | Prop=Normal { span: Span { lo: BytePos(276), hi: BytePos(281), ctxt: #0 }, sym: Atom('value' type=inline) }
  | NoSuchProperty {
  |     span: Span {
  |         lo: BytePos(
  |             271,
  |         ),
  |         hi: BytePos(
  |             281,
  |         ),
  |         ctxt: #0,
  |     },
  |     obj: Some(
  |         Arc(
  |             Freezed {
  |                 ty: Interface(
  |                     Interface {
  |                         span: Span {
  |                             lo: BytePos(
  |                                 271,
  |                             ),
  |                             hi: BytePos(
  |                                 275,
  |                             ),
  |                             ctxt: #0,
  |                         },
  |                         name: Base#6917,
  |                         type_params: None,
  |                         extends: [],
  |                         body: [
  |                             Property(
  |                                 PropertySignature {
  |                                     span: Span {
  |                                         lo: BytePos(
  |                                             24,
  |                                         ),
  |                                         hi: BytePos(
  |                                             47,
  |                                         ),
  |                                         ctxt: #0,
  |                                     },
  |                                     accessibility: None,
  |                                     readonly: true,
  |                                     key: Normal {
  |                                         span: Span {
  |                                             lo: BytePos(
  |                                                 33,
  |                                             ),
  |                                             hi: BytePos(
  |                                                 38,
  |                                             ),
  |                                             ctxt: #0,
  |                                         },
  |                                         sym: Atom('value' type=inline),
  |                                     },
  |                                     optional: false,
  |                                     params: [],
  |                                     type_ann: Some(
  |                                         Keyword(
  |                                             KeywordType {
  |                                                 span: Span {
  |                                                     lo: BytePos(
  |                                                         40,
  |                                                     ),
  |                                                     hi: BytePos(
  |                                                         46,
  |                                                     ),
  |                                                     ctxt: #0,
  |                                                 },
  |                                                 kind: TsNumberKeyword,
  |                                                 metadata: KeywordTypeMetadata {
  |                                                     common: CommonTypeMetadata {
  |                                                         implicit: false,
  |                                                         infected_by_this_in_object_literal: false,
  |                                                         prevent_converting_to_children: false,
  |                                                         no_expand: false,
  |                                                         ignore_no_expand: false,
  |                                                         contains_infer_type: false,
  |                                                         prevent_complex_simplification: false,
  |                                                         resolved_from_var: false,
  |                                                         prevent_generalization: false,
  |                                                         prevent_tuple_to_array: false,
  |                                                     },
  |                                                 },
  |                                             },
  |                                         ),
  |                                     ),
  |                                     type_params: None,
  |                                     metadata: TypeElMetadata {
  |                                         has_default: false,
  |                                     },
  |                                     accessor: Accessor {
  |                                         getter: false,
  |                                         setter: false,
  |                                     },
  |                                 },
  |                             ),
  |                         ],
  |                         metadata: InterfaceMetadata {
  |                             common: CommonTypeMetadata {
  |                                 implicit: false,
  |                                 infected_by_this_in_object_literal: false,
  |                                 prevent_converting_to_children: false,
  |                                 no_expand: false,
  |                                 ignore_no_expand: false,
  |                                 contains_infer_type: false,
  |                                 prevent_complex_simplification: false,
  |                                 resolved_from_var: true,
  |                                 prevent_generalization: false,
  |                                 prevent_tuple_to_array: false,
  |                             },
  |                         },
  |                     },
  |                 ),
  |             },
  |         ),
  |     ),
  |     prop: Some(
  |         Normal {
  |             span: Span {
  |                 lo: BytePos(
  |                     276,
  |                 ),
  |                 hi: BytePos(
  |                     281,
  |                 ),
  |                 ctxt: #0,
  |             },
  |             sym: Atom('value' type=inline),
  |         },
  |     ),
  | }
    ,-[$DIR/tests/conformance/types/intersection/intersectionTypeReadonly.ts:17:1]
 17 | base.value = 12 // error, lhs can't be a readonly property
    : ^^^^^^^^^^
    `----
```

### PR DESCRIPTION
**Pull request by sunrabbit123**
**Description:**

Resolve IntersectionReduction

**BREAKING CHANGE:**

if rhs.isAny, check `to` is never
if `to` is Never and rhs.isAny, fail!

and

LValue is Intersection

1. represent `intersection_normalize`
2. Fix if property is `readonly keyword | keyword`, throw readonly error
3. Fix #377
--- 
same #398

change the PR target branch because my branch is ruined

### CODE REVIEW

File: crates/stc_ts_type_checker/tests/conformance/types/mapped/recursiveMappedTypes.ts.stats.rust-debug [hunk 1]
```diff
@@ -2,5 +2,5 @@ Stats {
     required_error: 9,
     matched_error: 0,
     extra_error: 0,
-    panic: 0,
+    panic: 1,
```
Comment (sunrabbit123) [inline crates/stc_ts_type_checker/tests/conformance/types/mapped/recursiveMappedTypes.ts.stats.rust-debug:5]: It's Overflow of Recursive Values Issue

I think that We will should has solution about put dejavu in normalize
There may be other ways than this.

https://github.com/dudykr/stc/blob/36454c34724a476623376e771e9a7703697af328/crates/stc_ts_file_analyzer/src/analyzer/types/mod.rs#L526-L533

**Review by kdy1 — CHANGES_REQUESTED** [review]
Can you investigate the CI failure?

**Comment by sunrabbit123** [comment]
I investigate Extra_error and panic added

**Review by sunrabbit123 — COMMENTED** [review]
For now, we've finished the investigation here.

File: crates/stc_ts_type_checker/tests/conformance/types/intersection/intersectionOfUnionNarrowing.ts.stats.rust-debug [hunk 2]
```diff
@@ -1,6 +1,6 @@
 Stats {
     required_error: 0,
     matched_error: 0,
-    extra_error: 0,
+    extra_error: 1,
```
Comment (sunrabbit123) [inline crates/stc_ts_type_checker/tests/conformance/types/intersection/intersectionOfUnionNarrowing.ts.stats.rust-debug:4]: `else narrowing` is incorrect

```ts
declare const a : ((undefined | {
    bProp: string;
}) & ( object)); // { bProp: string; } & object
declare const b : ((undefined | {
    bProp: string;
}) & ( undefined | object)); // ({ bProp: string;} & object) | undefined
```
It should be evaluated as A, but it is evaluated as B

File: crates/stc_ts_type_checker/tests/conformance/controlFlow/controlFlowIfStatement.ts.stats.rust-debug [hunk 3]
```diff
@@ -1,6 +1,6 @@
 Stats {
     required_error: 0,
     matched_error: 0,
-    extra_error: 0,
+    extra_error: 1,
```
Comment (sunrabbit123) [inline crates/stc_ts_type_checker/tests/conformance/controlFlow/controlFlowIfStatement.ts.stats.rust-debug:4]: this too,
wrong narrowing

**Review by sunrabbit123 — COMMENTED** [review]
Where is the narrow due to the if?

**Comment by kdy1** [comment]
@sunrabbit123 It depends on the kind of narrowing, but you can search for `fn narrow_` using IDE

**Comment by sunrabbit123** [comment]
I don't know which narrowing to use.

I found narrow
1. narrow_with_predicate
2. narrow_with_equality

I didn't see anything else.

Also, I didn't find any syntax for putting type in else.

**Comment by kdy1** [comment]
image

For `false` parts, please refer to usage sides of those functions

File: crates/stc_ts_type_checker/tests/conformance/types/mapped/mappedTypeRelationships.ts.stats.rust-debug [hunk 4]
```diff
@@ -1,6 +1,6 @@
 Stats {
-    required_error: 27,
-    matched_error: 1,
-    extra_error: 25,
-    panic: 0,
```
Comment (sunrabbit123) [inline crates/stc_ts_type_checker/tests/conformance/types/mapped/mappedTypeRelationships.ts.stats.rust-debug:5]: `IndexedAccessType` cause overflow,
if you want decrase panic, you should overflow narrow any

File: crates/stc_ts_type_checker/tests/conformance/types/mapped/mappedTypeConstraints2.ts.stats.rust-debug [hunk 5]
```diff
@@ -1,6 +1,6 @@
 Stats {
-    required_error: 1,
-    matched_error: 2,
-    extra_error: 2,
-    panic: 0,
+    required_error: 3,
+    matched_error: 0,
+    extra_error: 0,
+    panic: 1,
```
Comment (sunrabbit123) [inline crates/stc_ts_type_checker/tests/conformance/types/mapped/mappedTypeConstraints2.ts.stats.rust-debug:5]: this too,

File: crates/stc_ts_type_checker/tests/conformance/types/literal/numericStringLiteralTypes.ts.stats.rust-debug [hunk 6]
```diff
@@ -2,5 +2,5 @@ Stats {
     required_error: 0,
     matched_error: 0,
     extra_error: 0,
-    panic: 0,
+    panic: 1,
```
Comment (sunrabbit123) [inline crates/stc_ts_type_checker/tests/conformance/types/literal/numericStringLiteralTypes.ts.stats.rust-debug:5]: this too

**Comment by sunrabbit123** [comment]
thx, Now I'm going to start working on Narrow.

File: crates/stc_ts_file_analyzer/tests/pass/types/intersection/ofUnionNarrowing/1.swc-stderr [hunk 7]
```diff
@@ -0,0 +1,108 @@
+
+  x Type
+    ,-[$DIR/tests/pass/types/intersection/ofUnionNarrowing/1.ts:11:1]
+ 11 | if (q.a !== undefined) {
+    :     ^
+    `----
+
+Error: 
+  > (({
```
Comment (sunrabbit123) [inline crates/stc_ts_file_analyzer/tests/pass/types/intersection/ofUnionNarrowing/1.swc-stderr:9]: X & ({
  |     a: object;
  |     b: undefined;
  | } | {
  |     a: undefined;
  |     b: object;
  | })

File: crates/stc_ts_file_analyzer/tests/pass/types/intersection/ofUnionNarrowing/1.swc-stderr [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+  | } & object)
+
+  x Type
+    ,-[$DIR/tests/pass/types/intersection/ofUnionNarrowing/1.ts:15:5]
+ 15 | q.b.bProp;
+    : ^^^^^^^^^
+    `----
+
+Error: 
+  > any
```
Comment (sunrabbit123) [inline crates/stc_ts_file_analyzer/tests/pass/types/intersection/ofUnionNarrowing/1.swc-stderr:108]: string

**Review by sunrabbit123 — COMMENTED** [review]
type is wrong

File: crates/stc_ts_file_analyzer/tests/pass/types/mapped/recursiveMappedTypes/3.swc-stderr [hunk 9]
```diff
@@ -0,0 +1,9 @@
+
+  x Type
+    ,-[$DIR/tests/pass/types/mapped/recursiveMappedTypes/3.ts:16:1]
+ 16 | a // "type" | "minimum_count" | "maximum_count" | undefined
+    : ^
+    `----
+
+Error: 
+  > ("type" | "minimum_count" | "maximum_count" | "collapsable")
```
Comment (sunrabbit123) [inline crates/stc_ts_file_analyzer/tests/pass/types/mapped/recursiveMappedTypes/3.swc-stderr:9]: this should bacome `never`

File: crates/stc_ts_file_analyzer/tests/pass/controlFlow/guard/assertions/9.swc-stderr [hunk 10]
```diff
@@ -46,7 +46,7 @@ Error:
     `----
 
 Error: 
-  > (string | undefined | null)
+  > string
```
Comment (sunrabbit123) [inline crates/stc_ts_file_analyzer/tests/pass/controlFlow/guard/assertions/9.swc-stderr:49]: can't disappear.

File: crates/stc_ts_file_analyzer/tests/pass/exprs/typeAssertion/3.swc-stderr [hunk 11]
```diff
@@ -42,4 +42,4 @@ Error:
     `----
 
 Error: 
-  > (string | number | undefined)
+  > (string | number)
```
Comment (sunrabbit123) [inline crates/stc_ts_file_analyzer/tests/pass/exprs/typeAssertion/3.swc-stderr:45]: can't disappear.ㅍ

File: crates/stc_ts_file_analyzer/tests/pass/exprs/typeAssertion/4.swc-stderr [hunk 12]
```diff
@@ -60,4 +60,4 @@ Error:
     `----
 
 Error: 
-  > (string | number | undefined)
+  > (string | number)
```
Comment (sunrabbit123) [inline crates/stc_ts_file_analyzer/tests/pass/exprs/typeAssertion/4.swc-stderr:63]: can't disappear.

File: crates/stc_ts_file_analyzer/tests/pass/exprs/typeAssertion/8.swc-stderr [hunk 13]
```diff
@@ -6,7 +6,7 @@
    `----
 
 Error: 
-  > (string | undefined | null)
+  > string
```
Comment (sunrabbit123) [inline crates/stc_ts_file_analyzer/tests/pass/exprs/typeAssertion/8.swc-stderr:9]: can't disappear.

File: crates/stc_ts_file_analyzer/tests/pass/types/intersection/intersectionWithUnionConstraint/1.swc-stderr [hunk 14]
```diff
@@ -6,4 +6,4 @@
    `----
 
 Error: 
-  > (T & U)
+  > string
```
Comment (sunrabbit123) [inline crates/stc_ts_file_analyzer/tests/pass/types/intersection/intersectionWithUnionConstraint/1.swc-stderr:9]: can't disappear.

File: crates/stc_ts_file_analyzer/tests/pass/types/intersection/intersectionWithUnionConstraint/2.swc-stderr [hunk 15]
```diff
@@ -6,4 +6,4 @@
    `----
 
 Error: 
-  > (T & U)
+  > string
```
Comment (sunrabbit123) [inline crates/stc_ts_file_analyzer/tests/pass/types/intersection/intersectionWithUnionConstraint/2.swc-stderr:9]: can't disappear.

File: crates/stc_ts_file_analyzer/tests/pass/types/intersection/intersectionWithUnionConstraint/3.swc-stderr [hunk 16]
```diff
@@ -6,4 +6,4 @@
    `----
 
 Error: 
-  > (T & U)
+  > string
```
Comment (sunrabbit123) [inline crates/stc_ts_file_analyzer/tests/pass/types/intersection/intersectionWithUnionConstraint/3.swc-stderr:9]: can't disappear.

File: crates/stc_ts_file_analyzer/tests/pass/controlFlow/guard/assertions/7.swc-stderr [hunk 17]
```diff
@@ -55,7 +55,7 @@ Error:
     `----
 
 Error: 
-  > (string | number | undefined)
+  > (string | number)
```
Comment (sunrabbit123) [inline crates/stc_ts_file_analyzer/tests/pass/controlFlow/guard/assertions/7.swc-stderr:58]: can't disappear.

File: crates/stc_ts_file_analyzer/tests/pass/controlFlow/guard/assertions/6.swc-stderr [hunk 18]
```diff
@@ -55,7 +55,7 @@ Error:
     `----
 
 Error: 
-  > (string | number | undefined)
+  > (string | number)
```
Comment (sunrabbit123) [inline crates/stc_ts_file_analyzer/tests/pass/controlFlow/guard/assertions/6.swc-stderr:58]: can't disappear.

File: crates/stc_ts_file_analyzer/tests/pass/controlFlow/guard/assertions/5.swc-stderr [hunk 19]
```diff
@@ -55,7 +55,7 @@ Error:
     `----
 
 Error: 
-  > (string | number | undefined)
+  > (string | number)
```
Comment (sunrabbit123) [inline crates/stc_ts_file_analyzer/tests/pass/controlFlow/guard/assertions/5.swc-stderr:58]: can't disappear.

File: crates/stc_ts_file_analyzer/tests/pass/controlFlow/guard/assertions/4.swc-stderr [hunk 20]
```diff
@@ -55,7 +55,7 @@ Error:
     `----
 
 Error: 
-  > (string | number | undefined)
+  > (string | number)
```
Comment (sunrabbit123) [inline crates/stc_ts_file_analyzer/tests/pass/controlFlow/guard/assertions/4.swc-stderr:58]: can't disappear.

**Review by sunrabbit123 — COMMENTED** [review]
why don't apply strictNullCheck Options

**Pull request closed without merging**